packages/flutter/lib/src/material/date.dart

```diff
@@ -244,4 +246,80 @@ class DateTimeRange {
 
   @override
   String toString() => '$start - $end';
+}                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                
+
+/// Encapsulates a possible start and end [DateTime] that represent a range of dates.
+///
+/// The range may include the [start] and [end] dates. The [start] and [end] dates
+/// may be equal to indicate a date range of a single day.
+/// The [DateTimeRangeValue] is valid if [start] is not after [end]
+@immutable
+class DateTimeRangeValue {
+  /// Creates a date range for the given start and end [DateTime].
+  DateTimeRangeValue({
+    DateTime? start,
+    DateTime? end,
+  })  : assert((start == null && end == null) ||
+            (start != null && end == null) ||
+            (start != null && !end!.isBefore(start))),
+        start = start != null ? DateUtils.dateOnly(start) : null,
+        end = end != null ? DateUtils.dateOnly(end) : null;
+
+  /// Creates an empty date range where start and end are null.
+  const DateTimeRangeValue.empty()
+      : start = null,
+        end = null;
+
+  /// The start of the range of dates.
+  final DateTime? start;
+
+  /// The end of the range of dates.
+  final DateTime? end;
+
+  @override
+  bool operator ==(Object other) {
+    if (other.runtimeType != runtimeType) {
+      return false;
+    }
+    return other is DateTimeRange && other.start == start && other.end == end;
+  }
+
+  @override
+  int get hashCode => Object.hash(start, end);
+
+  @override
+  String toString() => 'DateRangeValue($start - $end)';
+
+  DateTimeRangeValue copyWith({
+    DateTime? start,
+    DateTime? end,
+  }) {
+    return DateTimeRangeValue(
+      start: start != null ? DateUtils.dateOnly(start) : this.start,
+      end: end != null ? DateUtils.dateOnly(end) : this.end,
+    );
+  }
+
+  Map<String, dynamic> toMap() {
+    return <String, dynamic>{
+      'start': start?.millisecondsSinceEpoch,
+      'end': end?.millisecondsSinceEpoch,
+    };
+  }
+
+  factory DateTimeRangeValue.fromMap(Map<String, dynamic> map) {
+    return DateTimeRangeValue(
+      start: map['start'] != null
+          ? DateTime.fromMillisecondsSinceEpoch(map['start'] as int)
+          : null,
+      end: map['end'] != null
+          ? DateTime.fromMillisecondsSinceEpoch(map['end'] as int)
+          : null,
+    );
+  }
+
+  String toJson() => json.encode(toMap());
```

Comment:
in general if we can avoid using JSON that would be ideal; what's the JSON used for here?

Reply:
RestorableProperty has a toPrimitive and fromPrimitive. Before it was two separate RestorableDateTime properties. Now It's one separate RestorableChangeController holding a `DateTimeRangeValue`